add initial implementation of CUSTOM action in AuthorizationPolicy

yangminzhu opened the pull request:
This PR implements the CUSTOM action and tested manually for now, remaining work:
1. Fix to use the namespace of MeshConfig
2. Fix to properly support k8s service and ServiceEntry
3. E2E tests with sample ext_authz server



[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure


Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any changes that may affect Istio users.

istio-testing commented:
Skipping CI for Draft Pull Request.
If you want CI signal for your change, please convert it to an actual PR.
You can still manually trigger a test run with `/test all`

yangminzhu commented:
/test all

istio-testing commented:
@yangminzhu: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 442a61a391fc97cfdb0f2f42721257a1277f4b0f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/unit-tests_istio/1322334154716614656) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@yangminzhu: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 442a61a391fc97cfdb0f2f42721257a1277f4b0f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/unit-tests_istio/1322334154716614656) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | 442a61a391fc97cfdb0f2f42721257a1277f4b0f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-pilot-k8s-tests_istio/1322334154762752000) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@yangminzhu: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 442a61a391fc97cfdb0f2f42721257a1277f4b0f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/unit-tests_istio/1322334154716614656) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | 442a61a391fc97cfdb0f2f42721257a1277f4b0f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-pilot-k8s-tests_istio/1322334154762752000) | `/test integ-pilot-k8s-tests_istio`
integ-security-k8s-tests_istio | 442a61a391fc97cfdb0f2f42721257a1277f4b0f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-security-k8s-tests_istio/1322334154792112128) | `/test integ-security-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

yangminzhu commented:
/test all

istio-testing commented:
@yangminzhu: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 442a61a391fc97cfdb0f2f42721257a1277f4b0f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/unit-tests_istio/1322334154716614656) | `/test unit-tests_istio`
integ-security-k8s-tests_istio | 442a61a391fc97cfdb0f2f42721257a1277f4b0f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-security-k8s-tests_istio/1322334154792112128) | `/test integ-security-k8s-tests_istio`
integ-pilot-k8s-tests_istio | a1e9ca92f02f7bdd717460b8562341269157bad4 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-pilot-k8s-tests_istio/1322358663217680384) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

yangminzhu commented:
/test integ-pilot-k8s-tests_istio

istio-testing commented:
@yangminzhu: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-k8s-tests_istio | 442a61a391fc97cfdb0f2f42721257a1277f4b0f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-security-k8s-tests_istio/1322334154792112128) | `/test integ-security-k8s-tests_istio`
integ-multicluster-k8s-tests_istio | a1e9ca92f02f7bdd717460b8562341269157bad4 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-multicluster-k8s-tests_istio/1322358663288983552) | `/test integ-multicluster-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@yangminzhu: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | a1e9ca92f02f7bdd717460b8562341269157bad4 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-multicluster-k8s-tests_istio/1322358663288983552) | `/test integ-multicluster-k8s-tests_istio`
integ-ipv6-k8s-tests_istio | 32b561095c721652a1e6a53af7d848537a3133a0 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-ipv6-k8s-tests_istio/1323370841429250048) | `/test integ-ipv6-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

yangminzhu commented:
/retest

istio-testing commented:
@yangminzhu: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | a18b9dcf42700af15c962c6476c9639ffb92a178 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-multicluster-k8s-tests_istio/1323479717122150400) | `/test integ-multicluster-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@yangminzhu: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | a18b9dcf42700af15c962c6476c9639ffb92a178 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-multicluster-k8s-tests_istio/1323479717122150400) | `/test integ-multicluster-k8s-tests_istio`
integ-security-k8s-tests_istio | a18b9dcf42700af15c962c6476c9639ffb92a178 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-security-k8s-tests_istio/1323479717071818752) | `/test integ-security-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

yangminzhu commented:
/retest

incfly commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,315 @@
+
+func buildExtAuthz(configs []*meshconfig.MeshConfig_ExtensionProvider, providers []string) (*builtExtAuthz, error) {
+	if hasDuplicate(providers) {
+		return nil, fmt.Errorf("only 1 extension provider is allowed for a workload, found different providers: %v", providers)
```
If seems like the `hasDuplicate` is acutally `notAllTheSame`, right? The function name is confusing.

yangminzhu replied:
ah, yes, updated the function name and reworded the error message.

incfly commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,315 @@
+
+func buildExtAuthz(configs []*meshconfig.MeshConfig_ExtensionProvider, providers []string) (*builtExtAuthz, error) {
+	if hasDuplicate(providers) {
+		return nil, fmt.Errorf("only 1 extension provider is allowed for a workload, found different providers: %v", providers)
```
The condition is checking whether names are duplicated, but the error msg is about "only 1 allowed"?

yangminzhu replied:
see above.

incfly commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,315 @@
+		return nil, errs
+	}
+
+	authzLog.Debugf("resolved provider %s to config: %v", provider, spew.Sdump(ret))
```
Capitalizae log line

incfly commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,315 @@
+		namespace, name = "istio-system", parts[0]
+	}
+
+	// TODO(yangminzhu): This is temporary and still under development.
```
Is this comment outdated?

incfly replied:
Or you want to address in follow-up PR? If so, would need reword

yangminzhu replied:
will be addressed in follow PR, updated comment.

incfly commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,315 @@
+	return hostname, cluster, nil
+}
+
+func parsePort(port uint32) (int, error) {
```
Are there some model package util function for this kinda definition? I feel very likely

yangminzhu replied:
there is one in validation pkg but it caused some cyclic import issue.

incfly commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,315 @@
+	service := &extauthzhttp.HttpService{
+		PathPrefix: config.PathPrefix,
+		ServerUri: &envoy_config_core_v3.HttpUri{
+			// Timeout is required, use a large value as a placeholder, the real timeout is controlled in DestinationRule.
```
why it's required though? Just because of Envoy ext_authz config validation semantics? And this config does not take effect when we are using the Cluster generated by Istiod? If so, could you comment reason a bit?

yangminzhu replied:
required due to the validation semantic, the smaller timeout should take effect, updated comment.

incfly commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,315 @@
+		ServerUri: &envoy_config_core_v3.HttpUri{
+			// Timeout is required, use a large value as a placeholder, the real timeout is controlled in DestinationRule.
+			Timeout: &duration.Duration{Seconds: 600},
+			// Uri is required but actually not used in the ext_authz filter.
```
Sounds like ext authz filter's wrong constraint as well, as above?

(no AI, just confirm)

yangminzhu replied:
ext_authz filter is using referring to the generic HttpUri API, not only validation makes sense in the ext_authz context.

incfly replied:
because envoy is validating as HTTP URI typed, i see.

"s/required/required for Envoy's validation/g" optional

incfly commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,315 @@
+
+func generateGRPCConfig(cluster string, failOpen bool, status *envoytypev3.HttpStatus) *builtExtAuthz {
+	// The cluster includes the character `|` that is invalid in gRPC authority header and will cause the connection
+	// rejected in the server side, replace it with a valid character.
```
If we replace the cluster name's character, does that require us to change the cds to duplicate the coresponding cluster and replace name with `-` as well? Or there're some magic with the replacing?

yangminzhu replied:
this is just replacing the authority used in the ext_authz config, the cluster name is not changed.

incfly commented on pkg/config/validation/extensionprovider.go:
```diff
@@ -0,0 +1,120 @@
+	}
+	code, err := strconv.ParseInt(status, 10, 32)
+	if err != nil {
+		return fmt.Errorf("invalid statusOnError value %s: %v", status, err)
```
The func is passing a `status` string, without knowing it's actually passing the `x.status_on_error` field. the function naming seems is leaking details how how it should be used.

yangminzhu replied:
I think it's fine, this is a local function (not exported) and is only supposed to be used in this package.

incfly commented on pilot/pkg/security/authz/builder/builder.go:
```diff
@@ -143,30 +190,83 @@ func build(policies []model.AuthorizationPolicy, tdBundle trustdomain.Bundle,
+			},
+		}
+	}
+	// Add the RBAC filter in shadow mode so that it only evaluates the matching rules for CUSTOM action but not enforce it.
```
I read the sample output and code and here is my understanding: 

- we generate additional rbac filter, with shadow deny policies. Those rbac filter's condition is based on ext_authz condition.
- Then the ext authz filter is using the preivous rbac filte's output metadata filter, with `filterEnabledMetadata` to determine whether acutal kick in.

In some way, the first rbac is only to prepapre the triggering condition for the ext authz's config limitation?

Is that correct? Feel free to point to some section in the design doc. I do feel though, this is worth a bit more detailed comments around the code, maybe here and other places.

yangminzhu replied:
that's correct, I think this is already pretty clear with the current comment, added a link to the design doc for more details.

incfly replied:
Where's the link of the design doc? didn't see it nearby here.

We can add one sentence like "...keyed by the policy. And then the ext_authz filter can utilize these metadata to trigger the enforcement conditionally."

optional.

yangminzhu replied:
updated the comment, the design link is just below.

incfly commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,315 @@
+)
+
+const (
+	extAuthzMatchPrefix = "istio-ext-authz"
```
are you aware of any tooling(istioctl) or something is making dependency on the format of this rule's name? just seeing this is adding `-` in between in a slight different way than before.

Fine if it's totally internal.

yangminzhu replied:
this should be totally internal.

yangminzhu commented on pilot/pkg/networking/plugin/plugin.go:
```diff
@@ -20,9 +20,11 @@ import (
 )
 
 const (
+	// AuthzCustom is the name of the authorization plugin (CUSTOM action) passed through the command line
```
this is the common line argument in istiod, a global flag that controls the authn/z plugin itself, honestly I think we don't need this and should just enable authn/z plugins by default.

yangminzhu commented:
@incfly thanks for the review, updated PR to address comments so far.

Also @istio/wg-networking-maintainers-pilot can someone take a look from the networking side? thanks.

liminw commented on pilot/pkg/security/authz/builder/builder.go:
```diff
@@ -39,70 +42,108 @@ var (
 func (b Builder) BuildTCP() []*tcppb.Filter {
-	var filters []*tcppb.Filter
+	if b.option.IsCustomBuilder {
+		if configs := b.build(b.customPolicies, rbacpb.RBAC_DENY, true); configs != nil {
```
Why is custom builder using DENY filter?

yangminzhu replied:
DENY will cause an invalid config (e.g. found HTTP attribute in TCP filter chain) to always trigger the action (calling ext_authz), it doesn't matter in runtime as it's shadow mode (only evaluate but not enforce).

liminw replied:
I see. Thanks.

yangminzhu commented:
@costinm @howardjohn could you help take a look? thank you.

yangminzhu commented on pilot/pkg/networking/plugin/authz/authorization.go:
```diff
@@ -30,45 +29,56 @@ var (
+
+const (
+	// Local for action ALLOW, DENY and AUDIT and is enforced by Envoy RBAC filter.
+	Local ActionType = iota
```
it's used to distinguish the CUSTOM action from other actions (ALLOW/DENY/AUDIT), the CUSTOM action is handled separately and added to the beginning of the filter chain.

hzxuzhonghu commented on pilot/pkg/security/authz/builder/builder.go:
```diff
@@ -112,61 +153,122 @@ func build(policies []model.AuthorizationPolicy, tdBundle trustdomain.Bundle,
+	return []*httppb.HttpFilter{
+		{
+			Name:       authzmodel.RBACHTTPFilterName,
+			ConfigType: &httppb.HttpFilter_TypedConfig{TypedConfig: util.MessageToAny(rbac)},
```
Does this mean this filter will be executed twice?

yangminzhu replied:
It depends on your authz policy. If you have Authz policy with both CUSTOM and non-CUSTOM action, then yes. The 1st RBAC will only evaluate the matching rule. We already do this for DENY policy (it will be translated to a different RBAC filter).

hzxuzhonghu commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,317 @@
+		errs = multierror.Append(errs, fmt.Errorf("at least 1 extension provider must be defined"))
+	}
+	for i, config := range configs {
+		if config.Name == "" {
+			errs = multierror.Append(errs, fmt.Errorf("extension provider name must not be empty, found empty at index: %d", i))
+		} else if _, found := resolved[config.Name]; found {
+			errs = multierror.Append(errs, fmt.Errorf("extension provider name must be unique, found duplicate: %s", config.Name))
```
This should be validated during mesh configuration loading

yangminzhu replied:
how? This needs the list of authz policies applied for a given workload, we don't know the workload in mesh configuration loading. Also I'm not sure how useful this could be, it seems the final result is mostly the same regardless where you do the validation.

yangminzhu replied:
oh, I looked at the wrong line (70), but actually this is also validated in mesh configuration loading in pkg/config/validation/extensionprovider.go

hzxuzhonghu replied:
I am still confused about why we need to duplicate validating it.

yangminzhu replied:
there are some validation that we can only do here because we don't know the workload and policies applied to it in ValidateMeshConfig().

hzxuzhonghu replied:
make sense

hzxuzhonghu commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,317 @@
+		var err error
+		switch p := config.Provider.(type) {
+		case *meshconfig.MeshConfig_ExtensionProvider_EnvoyExtAuthzHttp:
+			parsed, err = buildExtAuthzHTTP(p.EnvoyExtAuthzHttp)
```
Can we cache the result

yangminzhu replied:
probably in a follow PR, there might still some big changes to the implementation here, added a TODO.

istio-testing commented:
@yangminzhu: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-k8s-tests_istio | 6db3eae7e30e5fa6a9de3719a53710bb6d7f828b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-security-k8s-tests_istio/1324406078808526848) | `/test integ-security-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

yangminzhu commented:
/test integ-security-k8s-tests_istio

howardjohn commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,318 @@
+
+	// TODO(yangminzhu): The following is temporary and still under development, the next PR is to to properly get the
+	// corresponding cluster name for the k8s service and ServiceEntry.
+	hostname = fmt.Sprintf("%s.%s.svc.cluster.local", name, namespace)
```
do not use cluster.local, we support custom domains

yangminzhu replied:
yes, this is temporary and still under development (see comment above), I will fix this part in a follow up PR, it will touch more networking code and we will also likely get the hostname from the push context (e.g. serviceIndex).

howardjohn commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,318 @@
+	}
+	parts := strings.Split(service, "/")
+	if len(parts) > 2 {
+		return "", "", fmt.Errorf("service not in format [<namespace>/]<service>, found: %s", service)
```
Why is this namespace/service and not namespace/FQDN? doesn't that break ability to use ServiceEntry?

yangminzhu replied:
good question, I think we're trying to follow with the more general convention to refer to other (k8s or Istio) resource by using the name directly, and the indirection gives more flexibility (e.g. the user doesn't need to specify the cluster.local part)

This works well with k8s service but not with ServiceEntry because you can have multiple hosts. I think we will probably support only 1 host in ServiceEntry if it is being used in extension provider here. cc @louiscryan

howardjohn commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,318 @@
+	service := &extauthzhttp.HttpService{
+		PathPrefix: config.PathPrefix,
+		ServerUri: &envoy_config_core_v3.HttpUri{
+			// Timeout is required. Use a large value as a placeholder and so that the timeout in DestinationRule (should
```
timeout in DR is connect timeout, not http request timeout

yangminzhu replied:
Good point, We currently don't expose this timeout in the API so we have to use some default value here, added a TODO to revisit this and investigate if we need to expose it later.

By the way, the http request timeout here is `measured since when the frame with end_stream flag is sent until when the first frame is received`. For ext_authz request (not stream) that is probably close to the connect timeout in DR?

howardjohn commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,318 @@
+	// The cluster includes the character `|` that is invalid in gRPC authority header and will cause the connection
+	// rejected in the server side, replace it with a valid character and set in authority otherwise ext_authz will
+	// use the cluster name as default authority.
+	authority := strings.ReplaceAll(cluster, "|", "-")
```
We already have a header compatible, BuildDNSSrvSubsetKey. Should we use this?

yangminzhu replied:
probably not, I think this is only used in `sni` but not the actual cluster name?

howardjohn replied:
Why not? I don't want to invent a new format. BuildDNSSrvSubsetKey is meant for _exactly_ this use case.

yangminzhu replied:
Do you mean to change the networking code to use BuildDNSSrvSubsetKey for the cluster name in the first place? If that is okay (not sure if it has any significant side effect),  that could be done as a separate refactor in the networking core code.

Here we are only referring to the already generated cluster name that still uses the `|` format.

howardjohn replied:
No, I mean instead of replacing `|` with `-` replace it with `_.` which is already the standardized format

yangminzhu replied:
Oh, I misunderstood your original comment, will do in a follow PR.

howardjohn reviewed (commented):
LGTM aside from small comments

yangminzhu commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,318 @@
+		return nil, errs
+	}
+
+	authzLog.Debugf("Resolved provider %s to config: %v", provider, spew.Sdump(ret))
```
AH, good point, updated.

yangminzhu commented on pilot/pkg/security/authz/builder/extauthz.go:
```diff
@@ -0,0 +1,318 @@
+			err = fmt.Errorf("unsupported extension provider: %s", config.Name)
+		}
+		if err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("failed to parse extension provider %s: %v", config.Name, err))
```
what do you mean `but ignore them`? We still return the errors if `errs != nil`. `extAuthz.http` should not be nil if `errs` is not nil, `extAuthz.tcp` might be but we will check it explicitly in builder.go.

howardjohn replied:
ack, misread the code

yangminzhu commented:
@howardjohn I addressed the comments so far, there are some issues that I have planned to fix in a upcoming PR (e.g. the cluster.local domain), would you mind to approve if this looks good to you now? thanks.

istio-testing commented:
@yangminzhu: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | c188caaa9bf377514ceafa79d2483d9729777763 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-multicluster-k8s-tests_istio/1324833863213518848) | `/test integ-multicluster-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@yangminzhu: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | c188caaa9bf377514ceafa79d2483d9729777763 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-multicluster-k8s-tests_istio/1324833863213518848) | `/test integ-multicluster-k8s-tests_istio`
integ-pilot-k8s-tests_istio | c188caaa9bf377514ceafa79d2483d9729777763 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-pilot-k8s-tests_istio/1324833863129632768) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

yangminzhu commented:
/retest

istio-testing commented:
@yangminzhu: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | c188caaa9bf377514ceafa79d2483d9729777763 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-multicluster-k8s-tests_istio/1324833863213518848) | `/test integ-multicluster-k8s-tests_istio`
integ-pilot-k8s-tests_istio | c188caaa9bf377514ceafa79d2483d9729777763 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/28451/integ-pilot-k8s-tests_istio/1324838269451505664) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

yangminzhu commented:
/test integ-pilot-k8s-tests_istio
